lib/ts/tokenlist.ts

```diff
@@ -0,0 +1,80 @@
+namespace Stacks {
+
+    /**
+     * A token list similar to DOMTokenList used for manipulating whitespace separated attributes.
+     * When this list is modified, the associated attribute will also be changed.
+     */
+    export class TokenList {
```

Comment:
This structure seems a little overkill to me. The majority of the time you're going to have a single value anyways and the rest of the time you just need to call `.split(' ')`. I don't foresee the need for `remove` either in any case that doesn't just involve blanking out the entire target attribute.

Reply:
You would be so... so.. surprised here.  We have produced the following scenarios **TWICE** in the last few months.

A button has a popover that announces its new functionality.  When the popover is dismissed it is discarded and the element's natural tooltip starts functioning as normal.  When the button is clicked, its action loads a new popover to be displayed immediately.

So, the button transitions through the following three states:

```html
1. <button data-controller="s-popover s-tooltip" aria-controls="intro" />
2. <button data-controller="s-tooltip" />
3. <button data-controller="s-tooltip s-popover" aria-controls="details" />
```

Even without the third step, we need a mechanism to cleanly remove `data-controller="s-popover" aria-controls="intro"` while preserving and hopefully being unaware of `data-controller="s-tooltip"`

Reply:
Are both situations with attaching and detaching a popover? I think you've mostly handled this in the new `attach/detachPopover` helpers you've added. Do you see this being used more often? This seems like a fairly unique situation where a user is dynamically attaching multiple controllers to a single element, which I feel shouldn't come up that often.

Reply:
Only one controller needs to be dynamically added or removed to introduce a risk of things breaking. (Again, the need to attach or remove a popover from an element that has another controller has come up several times in practice since we've added the popover controller.  It's far from a one-off experience and there are several places in code where we would have benefitted from this feature.)

Imagine for example, we have the following code `<button class="js-my-button" data-controller="s-popover">Some button</button>`. We need to remove the popover after an interaction and call `$('.js-my-button').removeAttr('data-controller');`.

Later the button gets a second controller and becomes `<button class="js-my-button" data-controller="s-popover my-button-controller">Some button</button>`. Now when we call `$('.js-my-button').removeAttr('data-controller');`, we remove two controllers instead of one.

Not having a standard surgical way of manipulating the controller list means we're accumulating complexity in our codebase that will lead to bugs and testing complexity.

Alternatives I considered were:
- Using the native `DOMTokenList` that is used for `Element.classList`.  This wasn't feasible because there is no public constructor.
- Using the token functionality that exists in Stimulus.  This wasn't feasible because Stimulus only listens to mutation events and doesn't surface anything to do the actual mutations.
- Having simple functions like `Stacks.addToControllerAttribute(element, value)`.  This would result in 6 to 9 methods and a lot of the same underlying complexity.
- Having simple functions like Stacks.addTokenToAttribute(element, 'data-controller', value)`.  This was also pretty verbose.

I ended up with this because it wasn't too verbose and it mirrored existing patterns in the DOM.

One step further would be to put helpers in `Element.prototype` so you could do `myElement.stimulusControllerList.remove('s-popover')` but that seemed like a bridge too far at the moment.